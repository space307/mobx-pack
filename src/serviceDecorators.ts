import { cloneDeep, last } from 'lodash';
import type {
  BindableEntity,
  BindableEntityConstructor,
  BinderConfig,
  Constructor,
  ServiceConfigCallbackSet,
} from './typing/common.js';

function validateName(name: string): boolean {
  return !!(name && typeof name === 'string' && /^[A-Za-z][A-Za-z0-9_]+$/.test(name));
}

function validateNameList(list: string[]): boolean {
  return list.reduce((acc, item) => {
    if (!validateName(item)) {
      acc = false;
    }
    return acc;
  }, true);
}

function createConfig(): BinderConfig {
  return {
    onStart: [],
    onStop: '',
    bindAs: '',
    onBind: [],
    onUnbind: [],
  };
}

function prepareConfig<T extends Constructor>(ctr: T): T & BindableEntityConstructor {
  const bindableCtr = ctr as unknown as T & BindableEntityConstructor;

  if (!bindableCtr.binderConfig) {
    bindableCtr.binderConfig = createConfig();
  } else {
    bindableCtr.binderConfig = cloneDeep(bindableCtr.binderConfig);
  }

  return bindableCtr;
}

function putServiceNamesToConfig(
  names: string[],
  ctr: BindableEntityConstructor,
  callbackName: string,
  optionName: keyof Pick<BinderConfig, 'onStart' | 'onBind' | 'onUnbind'>,
): void {
  if (names && names.length && callbackName) {
    names.forEach((serviceName: string) => {
      if (!validateName(serviceName)) {
        throw new Error(`Wrong service name "${serviceName}"
          passed to function "${callbackName}" decorator (service:${ctr.name ?? 'unknown'}).`);
      }
    });

    if (!ctr.binderConfig) {
      throw new Error(`binderConfig not found! (service:${ctr.name ?? 'unknown'}).`);
    }

    if (optionName === 'onBind' || optionName === 'onUnbind') {
      const option = ctr.binderConfig[optionName];

      if (option) {
        const existCallback = option.find(
          (callback: ServiceConfigCallbackSet): boolean => last(callback) === callbackName,
        );

        if (existCallback === undefined) {
          option.push([...names, callbackName]);
        } else {
          existCallback.splice(0, existCallback.length, ...names, callbackName);
        }
      }
    } else {
      ctr.binderConfig[optionName] = [...names, callbackName];
    }
  }
}

function putMethodNameToConfig(
  ctr: BindableEntityConstructor,
  callbackName: string,
  optionName: keyof Pick<BinderConfig, 'onStop'>,
): void {
  if (ctr.binderConfig) {
    ctr.binderConfig[optionName] = callbackName;
  }
}

export function bindAs(
  name: string,
): <T extends object>(ctr: Constructor<T>) => BindableEntityConstructor<T & BindableEntity> {
  if (typeof name === 'function') {
    throw new Error(
      `Wrong attributes passed to bindAs decorator (service:${(name as () => unknown).name}).`,
    );
  }
  return ctr => {
    if (!validateName(name)) {
      throw new Error(
        `Wrong name "${name}" passed to bindAs decorator (service:${ctr.name ?? 'unknown'}).`,
      );
    }

    const bindedCtr = prepareConfig(ctr);
    bindedCtr.binderConfig.bindAs = name;

    return bindedCtr;
  };
}

export function onBind<T extends object>(
  ...serviceNames: string[]
): (service: T, callbackName: string) => void {
  if (!serviceNames.length || !validateNameList(serviceNames)) {
    throw new Error(`Wrong attributes passed to onBind decorator (${serviceNames.join(',')}).`);
  }

  return (service, callbackName) => {
    const proto = service.constructor as BindableEntityConstructor;
    prepareConfig(proto);

    putServiceNamesToConfig(serviceNames, proto, callbackName, 'onBind');
  };
}

export function onUnbind<T extends object>(
  ...serviceNames: string[]
): (service: T, callbackName: string) => T {
  if (!serviceNames.length || !validateNameList(serviceNames)) {
    throw new Error(`Wrong attributes passed to onUnbind decorator (${serviceNames.join(',')}).`);
  }
  return (service, callbackName) => {
    const proto = service.constructor as BindableEntityConstructor;
    prepareConfig(proto);

    putServiceNamesToConfig(serviceNames, proto, callbackName, 'onUnbind');
    return service;
  };
}

export function onStart<T extends object>(
  ...serviceNames: string[]
): (service: T, callbackName: string) => T {
  if (!serviceNames.length || !validateNameList(serviceNames)) {
    throw new Error(`Wrong attributes passed to onStart decorator (${serviceNames.join(',')}).`);
  }
  return (service, callbackName) => {
    const proto = service.constructor as BindableEntityConstructor;
    prepareConfig(proto);

    putServiceNamesToConfig(serviceNames, proto, callbackName, 'onStart');
    return service;
  };
}

export function onStop<T extends object>(service: T, callbackName: string): void {
  const proto = service.constructor as BindableEntityConstructor;
  prepareConfig(proto);
  putMethodNameToConfig(proto, callbackName, 'onStop');
}
