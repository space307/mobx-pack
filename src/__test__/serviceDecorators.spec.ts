/* eslint-disable no-unused-vars,@typescript-eslint/no-unused-vars,@typescript-eslint/no-empty-function */
import { onStop, onStart, onUnbind, onBind, bindAs } from '../serviceDecorators';

describe('serviceDecorators test', () => {
  it('wrong bindAs param', () => {
    expect(() => {
      // @ts-expect-error Expected
      @bindAs
      // @ts-expect-error Expected
      class Service {}
    }).toThrow();

    expect(() => {
      // @ts-expect-error Expected
      @bindAs()
      // @ts-expect-error Expected
      class Service {}
    }).toThrow();

    expect(() => {
      // @ts-expect-error Expected
      @bindAs(1)
      // @ts-expect-error Expected
      class Service {}
    }).toThrow();
  });

  it('wrong onBind param', () => {
    expect(() => {
      @bindAs('test')
      // @ts-expect-error Expected
      class Service {
        // @ts-expect-error Expected
        @onBind
        onBind() {}
      }
    }).toThrow();

    expect(() => {
      @bindAs('test')
      // @ts-expect-error Expected
      class Service {
        @onBind()
        onBind() {}
      }
    }).toThrow();
    expect(() => {
      @bindAs('test')
      // @ts-expect-error Expected
      class Service {
        // @ts-expect-error Expected
        @onBind('test1', 1)
        onBind() {}
      }
    }).toThrow();
  });

  it('wrong onStart param', () => {
    expect(() => {
      @bindAs('test')
      // @ts-expect-error Expected
      class Service {
        // @ts-expect-error Expected
        @onStart
        onStart() {}
      }
    }).toThrow();

    expect(() => {
      @bindAs('test')
      // @ts-expect-error Expected
      class Service {
        @onStart()
        onStart() {}
      }
    }).toThrow();
    expect(() => {
      @bindAs('test')
      // @ts-expect-error Expected
      class Service {
        // @ts-expect-error Expected
        @onStart('test1', 1)
        onStart() {}
      }
    }).toThrow();
  });

  it('serviceDecorators config', () => {
    @bindAs('test')
    class Service {
      @onStart('init')
      onStart() {}

      @onBind('test1', 'test2')
      onBind() {}

      @onBind('test1', 'test2')
      onBindOnlyGarageService() {}

      @onUnbind('test1', 'test2')
      onUnbind() {}

      @onStop
      onStop() {}
    }

    // @ts-expect-error Expected
    expect(Service.binderConfig).toMatchSnapshot();
  });

  describe('extends class with decorators', () => {
    let ParentService: any;
    let ExtendedService: any;

    beforeEach(() => {
      @bindAs('ParentService')
      class ParentServiceClass {
        @onStart('onStartParentService')
        onStart() {}

        @onBind('onBindParentService1', 'onBindParentService2')
        onBind() {}

        @onBind('onBindParentService3', 'onBindParentService4')
        onBindAdditionServices() {}

        @onUnbind('onUnbindParentService1', 'onUnbindParentService2')
        onUnbind() {}

        @onUnbind('onUnbindParentService3', 'onUnbindParentService4')
        onUnbindAdditionServices() {}

        @onStop
        onStop() {}
      }

      ParentService = ParentServiceClass;
    });

    describe('bindAs', () => {
      beforeEach(() => {
        @bindAs('ExtendedService')
        class ExtendedServiceClass extends ParentService {}

        ExtendedService = ExtendedServiceClass;
      });

      it('should extends new class with other bindAs', () => {
        expect(ExtendedService.binderConfig.bindAs).toEqual('ExtendedService');
      });

      it('should stay ParentService without changes in bindAs after extend', () => {
        expect(ParentService.binderConfig.bindAs).toEqual('ParentService');
      });

      it('should extends new class with parent bindAs if not redefined', () => {
        class ExtendedServiceClass extends ParentService {}

        expect(ExtendedServiceClass.binderConfig.bindAs).toEqual('ParentService');
      });
    });

    describe('onStart', () => {
      beforeEach(() => {
        class ExtendedServiceClass extends ParentService {
          @onStart('onStartExtendedService')
          onStart() {}
        }

        ExtendedService = ExtendedServiceClass;
      });

      it('should extends new class with other onStart', () => {
        expect(ExtendedService.binderConfig.onStart).toEqual(['onStartExtendedService', 'onStart']);
      });

      it('should stay ParentService without changes in onStart after extend', () => {
        expect(ParentService.binderConfig.onStart).toEqual(['onStartParentService', 'onStart']);
      });

      it('should extends new class with parent onStart if not redefined', () => {
        class ExtendedServiceClass extends ParentService {}

        expect(ExtendedServiceClass.binderConfig.onStart).toEqual([
          'onStartParentService',
          'onStart',
        ]);
      });
    });

    describe('onBind', () => {
      beforeEach(() => {
        class ExtendedServiceClass extends ParentService {
          @onBind('onBindExtendedService1')
          onBind() {}

          @onBind('onBindExtendedService3')
          onBindAdditionExtendedServices() {}
        }

        ExtendedService = ExtendedServiceClass;
      });

      it('should extends new class with other onBind', () => {
        expect(ExtendedService.binderConfig.onBind).toEqual([
          ['onBindExtendedService1', 'onBind'],
          ['onBindParentService3', 'onBindParentService4', 'onBindAdditionServices'],
          ['onBindExtendedService3', 'onBindAdditionExtendedServices'],
        ]);
      });

      it('should stay ParentService without changes in onBind after extend', () => {
        expect(ParentService.binderConfig.onBind).toEqual([
          ['onBindParentService1', 'onBindParentService2', 'onBind'],
          ['onBindParentService3', 'onBindParentService4', 'onBindAdditionServices'],
        ]);
      });
    });

    describe('onUnbind', () => {
      beforeEach(() => {
        class ExtendedServiceClass extends ParentService {
          @onUnbind('onUnbindExtendedService1')
          onUnbind() {}

          @onUnbind('onUnbindExtendedService3')
          onUnbindAdditionExtendedServices() {}
        }

        ExtendedService = ExtendedServiceClass;
      });

      it('should extends new class with other onUnbind', () => {
        expect(ExtendedService.binderConfig.onUnbind).toEqual([
          ['onUnbindExtendedService1', 'onUnbind'],
          ['onUnbindParentService3', 'onUnbindParentService4', 'onUnbindAdditionServices'],
          ['onUnbindExtendedService3', 'onUnbindAdditionExtendedServices'],
        ]);
      });

      it('should stay ParentService without changes in onUnbind after extend', () => {
        expect(ParentService.binderConfig.onUnbind).toEqual([
          ['onUnbindParentService1', 'onUnbindParentService2', 'onUnbind'],
          ['onUnbindParentService3', 'onUnbindParentService4', 'onUnbindAdditionServices'],
        ]);
      });
    });

    describe('onStop', () => {
      beforeEach(() => {
        class ExtendedServiceClass extends ParentService {
          @onStop
          onStopExtended() {}
        }

        ExtendedService = ExtendedServiceClass;
      });

      it('should extends new class with other onStop', () => {
        expect(ExtendedService.binderConfig.onStop).toEqual('onStopExtended');
      });

      it('should stay ParentService without changes in onStop after extend', () => {
        expect(ParentService.binderConfig.onStop).toEqual('onStop');
      });

      it('should extends new class with parent onStop if not redefined', () => {
        class ExtendedServiceClass extends ParentService {}

        expect(ExtendedServiceClass.binderConfig.onStop).toEqual('onStop');
      });
    });
  });
});
