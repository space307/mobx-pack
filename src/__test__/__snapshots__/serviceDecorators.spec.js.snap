// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`serviceDecorators test serviceDecorators config 1`] = `
Object {
  "bindAs": "test",
  "onBind": Array [
    Array [
      "test1",
      "test2",
      "onBind",
    ],
    Array [
      "test1",
      "test2",
      "onBindOnlyGarageService",
    ],
  ],
  "onStart": Array [
    "init",
    "onStart",
  ],
  "onStop": "onStop",
  "onUnbind": Array [
    Array [
      "test1",
      "test2",
      "onUnbind",
    ],
  ],
}
`;
